### ISSUE TITLE
fork doesn't honor PATH setting

### ISSUE THREAD

**Issue by bill-torpey** [ForkIssues/Tracker#1889]
Because we care about file timestamps, we use a post-checkout hook that invokes `git-restore-mtime` (https://github.com/MestreLion/git-tools).  To simplify things, I add the script directory to my PATH in `.bash_profile`, but fork doesn't honor that setting.

**Comment by DanPristupov** [comment]
> To simplify things, I add the script directory to my PATH in .bash_profile, but fork doesn't honor that setting.

Then you must start Fork from bash, otherwise it won't receive that PATH (`open -a Fork`)
https://github.com/ForkIssues/Tracker/issues/430#issuecomment-429013890

**Comment by bill-torpey** [comment]
PATH is also set in `.bashrc`, but fork still can't find things.  Using a script to start fork is not great -- I already do that with SourceTree, which is one of the reasons I'm looking at fork.

**Comment by DanPristupov** [comment]
> PATH is also set in `.bashrc`, but fork still can't find things.

Fork doesn't read settings of some other applications. Same as other application don't read Fork settings.

In unix ENV is inherited from the parent process.

**Comment by bill-torpey** [comment]
> Fork doesn't read settings of some other applications. Same as other application don't read Fork settings.

Not the application, the OS.  Applications *are* expected to take settings from the OS.

> In unix ENV is inherited from the parent process.

Well, this is on Mac, so kinda unix, but not exactly.  Not sure how fork umm, forks, but at least on Linux there are options to preserve parent environment or not.  fork may have good reasons to not preserve the parent process' environment, but if so it would be very helpful to have a setting in fork itself to add directories to PATH/(DY)LD_LIBRARY_PATH.

**Comment by Andrew-Au** [comment]
> PATH is also set in `.bashrc`

Is the issue that you're setting the path in `.bashrc`, not `.profile` (or equivalent)?  The Finder session is not an interactive shell, so it will not source `.bashrc` when the session starts.

Fork is spawned by the Finder, so it also has no reason to run `.bashrc`.

Note: I think I'm mis-using the term Finder here, but the meaning should be obvious.

**Comment by bill-torpey** [comment]
Yes, it appears that macOS is rather strict about setting environment variables for GUI apps, which causes issues for a bunch of other apps as well, including SourceTree.

The latest release of SourceTree fixes this issue: https://jira.atlassian.com/browse/SRCTREE-7184. I can confirm that this works.  (FWIW, previous workaround was to invoke SourceTree from a command shell which was bundled as an app -- not great).

**Comment by Andrew-Au** [comment]
I wrote a little test app in Swift, and the environment passed to it via double-click in Finder was indeed minimal, without sourcing either `.zprofile` (login session) or `.zshrc` (interactive shell).

I'm going to go out on a limb and claim this is probably correct behaviour.  Although one could argue that the act of logging in via the Finder should trigger the profile.

However, the consequence is that any app that wants to run a Unix subprocess in the same context as a user shell needs to actually spawn the appropriate environment, not just run the executable in the host app's environment.

`$SHELL` is set, so you could try executing `{SHELL} command`.  Might need tweaks to ensure that the appropriate startup files are sourced (e.g. passing `-l` to the shell or adding a hyphen prefix to `argv[0]`).

**Comment by CodingMarkus** [comment]
Environment variables only inherit to child processes. If you set an environment variable in a process, only child processes will see it, not parent processes and also not processes on the same level. UI applications are no child processes of `bash` or `zsh`, unless you start them from such a shell.

The very first process that macOS starts after the kernel has loaded is `launchd`, which acts like `init` on Linux/FreeBSD and brings up the rest of the system. Everything launched down the pipe inherits the environment of `launchd` and passes that on to its children, unless the environment is overridden first, which Finder doesn't do. Actually, Finder does not even start applications itself, it uses Launch Services to do that and Launch Services instructs `launchd` to start the application (in previous versions two instances of `launchd` were running, a system one running as root and a user one running as the logged in user, the later one will start the UI application and provide the environment, but already in Monterey there's only one `launchd` instance left).

What you can do is to write your own process that first changes the environment and then launches Fork. E.g. start Automator app and create a new "Application". As action add `Run Shell Script`. Select `/bin/zsh` as shell and then add two lines:

```
export PATH=...
open -a Fork
```

Save it wherever you want and use that starter to start Fork. Note that no matter what shell you use in Automator, the shell will not read any config files (the shell is not launched in interactive mode or as login shell), the default PATH will be `/usr/bin:/bin:/usr/sbin:/sbin` in either case. So you would usually append to that whatever you need:

```
export PATH="$PATH:/whateverDir1:/whateverDir2:..."
```

**Comment by vrnvorona** [comment]
> Yes, it appears that macOS is rather strict about setting environment variables for GUI apps, which causes issues for a bunch of other apps as well, including SourceTree.
> 
> The latest release of SourceTree fixes this issue: [jira.atlassian.com/browse/SRCTREE-7184](https://jira.atlassian.com/browse/SRCTREE-7184). I can confirm that this works. (FWIW, previous workaround was to invoke SourceTree from a command shell which was bundled as an app -- not great).

If SourceTree does it, so can we? It's really annoying issue since 99% of tutorials and stuff teach to set up env in .zshrc and similar files and suddenly it doesn't work by itself.

**Comment by aparajita** [comment]
It's a MUST, otherwise commit hooks tend to fail.

Tower uses an `environment.plist` file in their Application Support directory.

**Comment by mnhattt** [comment]
I want to write macos automator script to open folk with selected forlder but I'm not success

`open -a 'fork' "$@"`

Can someone help.

**Comment by b2nkuu** [comment]
i found solution at is stack
https://stackoverflow.com/questions/51636338/what-does-launchctl-config-user-path-do

hope your help.

**Comment by chimit** [comment]
> Start Fork from your command line and it will derive ENV of the parent process. You can do that using open -a Fork command or just fork if you installed Fork CLI tools.
https://github.com/fork-dev/Tracker/issues/430#issuecomment-429013890

Installing Fork CLI tools doesn't fix the problem for me. Running Fork from the command line every time is not very convenient. Any chances to get it work out of the box, @DanPristupov?

**Comment by ValentinGurkov** [comment]
I am hoping some sort of a solution gets considered.
BTW, I belive more or less https://github.com/fork-dev/Tracker/issues/1227 also describes the same problem, probably the issues can be merged.

**Comment by wmadden** [comment]
@DanPristupov wouldn't this problem be solved by running git hooks in a new shell which can load the user's .profile etc itself according to their local setup? Nobody really cares what env vars Fork has access to, only the shell spawned to run the git hook commands

**Comment by niklasravnsborg** [comment]
@wmadden The git hook commands are probably executed directly without a shell; if it would spawn a shell, then this shell would take its own configuration into account, where the `PATH` could be modified.

For me personally, it would not only be important for git hooks but also for the `git` and `gpg` binaries, since I want to have control over which ones are used.

**Comment by wmadden** [comment]
I believe Fork has its own git implementation, no?

**Comment by niklasravnsborg** [comment]
> I believe Fork has its own git implementation, no?

You can customize the git binary in it's settings.

**Comment by DanPristupov** [comment]
I need a hand with testing. Can anyone contact me by email, please? support@fork.dev

**Comment by kjyv** [comment]
Any update about this? Otherwise I can help with testing. I'm unable to use anything in my pre-commit hooks that isn't installed system wide because something with System Shell does not seem to work (the open -a Fork workaround works but isn't nice)

**Comment by DanPristupov** [comment]
@kjyv enable 'System Shell' in Fork -> Preferences -> ENV Path.

**Comment by kjyv** [comment]
That's what I have but it does not seem to properly use the Path from there.
(it is: Integration -> ENV PATH -> System Shell '/bin/bash')

**Comment by DanPristupov** [comment]
@kjyv add `printenv` to your hook and run it from Fork and from the CLI

**Comment by kjyv** [comment]
Ok, so it seems the pre-commit hook gets a different env than the pre-push hook. Commiting does find the executable in ~/.local/bin/, pushing fails.

pre-commit: 
PATH=/Applications/Fork.app/Contents/Resources/git-instance/libexec/git-core:/Applications/Fork.app/Contents/Resources/git-instance/git-lfs:/Applications/Fork.app/Contents/Resources/gitflow-avh:/Users/stefan/.local/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/Users/stefan/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin:/Library/Frameworks/Mono.framework/Versions/Current/Commands

pre-push:
PATH=/Applications/Fork.app/Contents/Resources/git-instance/libexec/git-core:/Applications/Fork.app/Contents/Resources/git-instance/git-lfs:/Applications/Fork.app/Contents/Resources/gitflow-avh:/opt/homebrew/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin

**Comment by DanPristupov** [comment]
> it seems the pre-commit hook gets a different env than the pre-push hook

Yes, pre-push currently doesn't use this feature. But you asked about pre-commit hooks.

**Comment by kjyv** [comment]
Ah okay. I saw the failing push on [pre-commit](https://pre-commit.com/) hooks, which includes both, sorry for the confusion Do you plan to add the custom paths for pre-push as well?

**Comment by kjyv** [comment]
Sorry to ask again but is there any chance that this gets worked on some soon? It sounds like it would be a small change to also apply the existing functionality for pre-push. I currently can't use Fork for pushing unless I start if from the terminal...

**Comment by ivanggq** [comment]
I also would appreciate support for pre-push. Thanks for the amazing Fork client.

**Comment by ifeltsweet** [comment]
We just run into the same issue with pre-push.

**Comment by ValentinGurkov** [comment]
Wouldn't it be best to support the path on all git hook instead of selected few?

**Comment by kjyv** [comment]
This is still pretty annoying, @DanPristupov could you at least give some heads up if you will or will not fix this? Considering this is a paid product, I think a bit more feedback here is not too much to ask. I would be more patient or attempt to fix it myself if it was free or open source.

**Comment by DanPristupov** [comment]
@kjyv

> could you at least give some heads up if you will or will not fix this?

This will not be changed in the *next update*.

If the problem really bothers you, start Fork from the command line using `open -a Fork` and it will inherit ENV of the parent terminal process.

**Comment by arancauchi** [comment]
>If the problem really bothers you

It's kinda annoying

**Comment by haynar** [comment]
I found this quite annoying and came up with a workaround for macOS if anyone is interested.

1. Create an empty file and name it `Fork` (or anything else you like).
2. Make it executable by adding the respective file attribute:
    ```sh
    chmod +x ./Fork
    ```
3. Open it with any plain text editor and put this inside:
    ```sh
    #!/usr/bin/env sh
    open -a Fork
    exit
    ```

Congratulations, now you can double click on the new file to launch the Fork from terminal, making sure it has access to env variables. The drawback of this method is that it launches a new Terminal session to start the application and even if you make that window close automatically (see point 6) the Terminal app still remains open, I wasn't able to make it quit once the window is closed.

### Some optional steps to make everything look a bit nicer.

4. You can drag & drop the file into the right side of the dock (next to the Bin icon), but since it's not a real application it will not allow you to move it into application section.
5. To use the same icon from original Fork app open the file properties dialog of the original Fork.app by selecting the file in Finder and pressing `Cmd ⌘ + I` or right click and select "Get Info" from the context menu. Left click the application icon in the info dialog to select it, then copy it with `Cmd ⌘ + C`. Open the file properties dialog of the file created in the step 1, select the file icon and paste the icon with `Cmd ⌘ + V`. This will replace the default terminal icon of a shell script with the icon from Fork app. 
    <img width="539" height="562" alt="Image" src="https://github.com/user-attachments/assets/80489ec4-dff0-4506-b41c-99efdfc426b9" />
6. Since executing the shell script has a side effect of launching a new Terminal session, it will leave that window open after the script execution. To close the window automatically once the script is successfully executed go to `Terminal > Settings > Profiles > Shell` and select `Close if the shell exited cleanly` option for `When the shell exits`.
    <img width="672" height="628" alt="Image" src="https://github.com/user-attachments/assets/f0b6195d-5a2e-445f-a863-c1f3956898e4" />

**Comment by kristian-nst** [comment]
I asked AI if instead of using PATH, the commit hooks could try to find the absolute path to the used binaries

it came up with this eventually and it works for my case. Try to adjust to your/team's needs. YMMV:

```
if [ "$OS" = "Windows_NT" ]; then
	echo "Detected Windows"
else
	echo "Detected Unix/Linux or macOS"
	if command -v fnm >/dev/null 2>&1; then
		eval "$(fnm env)"
		fnm use $(cat .nvmrc) --log-level quiet
	fi
fi

BUN_CMD=""
if command -v bun >/dev/null 2>&1; then
	BUN_CMD="$(command -v bun)"
elif [ "$OS" = "Windows_NT" ] && command -v where.exe >/dev/null 2>&1; then
	CANDIDATE="$(where.exe bun 2>/dev/null | head -n 1 | tr -d '\r')"
	if [ -n "$CANDIDATE" ] && [ -x "$CANDIDATE" ]; then
		BUN_CMD="$CANDIDATE"
	fi
fi

if [ -z "$BUN_CMD" ] && [ -n "$BUN_INSTALL" ] && [ -x "$BUN_INSTALL/bin/bun" ]; then BUN_CMD="$BUN_INSTALL/bin/bun"; fi
if [ -z "$BUN_CMD" ] && [ -n "$BUN_INSTALL" ] && [ -x "$BUN_INSTALL/bin/bun.exe" ]; then BUN_CMD="$BUN_INSTALL/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -x "$HOME/.bun/bin/bun" ]; then BUN_CMD="$HOME/.bun/bin/bun"; fi
if [ -z "$BUN_CMD" ] && [ -x "$HOME/.bun/bin/bun.exe" ]; then BUN_CMD="$HOME/.bun/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -n "$USERPROFILE" ] && [ -x "$USERPROFILE/.bun/bin/bun.exe" ]; then BUN_CMD="$USERPROFILE/.bun/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -n "$LOCALAPPDATA" ] && [ -x "$LOCALAPPDATA/bun/bin/bun.exe" ]; then BUN_CMD="$LOCALAPPDATA/bun/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -n "$ChocolateyInstall" ] && [ -x "$ChocolateyInstall/bin/bun.exe" ]; then BUN_CMD="$ChocolateyInstall/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -x "$HOME/scoop/apps/bun/current/bun.exe" ]; then BUN_CMD="$HOME/scoop/apps/bun/current/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -x "/c/Users/$USERNAME/AppData/Local/bun/bin/bun.exe" ]; then BUN_CMD="/c/Users/$USERNAME/AppData/Local/bun/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -x "/c/ProgramData/chocolatey/bin/bun.exe" ]; then BUN_CMD="/c/ProgramData/chocolatey/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -x "/c/Program Files/bun/bin/bun.exe" ]; then BUN_CMD="/c/Program Files/bun/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -x "/c/Program Files (x86)/bun/bin/bun.exe" ]; then BUN_CMD="/c/Program Files (x86)/bun/bin/bun.exe"; fi
if [ -z "$BUN_CMD" ] && [ -x "/opt/homebrew/bin/bun" ]; then BUN_CMD="/opt/homebrew/bin/bun"; fi
if [ -z "$BUN_CMD" ] && [ -x "/usr/local/bin/bun" ]; then BUN_CMD="/usr/local/bin/bun"; fi

if [ -z "$BUN_CMD" ]; then
	echo "bun not found. Ensure Bun is installed and on PATH."
	exit 127
fi

echo "Bun command found at: $BUN_CMD"

# Ensure nested scripts that call `bun ...` can find it on PATH
BUN_DIR="$(dirname "$BUN_CMD")"
export PATH="$BUN_DIR:$PATH"

echo "Running pre-commit TSC..."
"$BUN_CMD" run tsc
echo "Running pre-commit Lint..."
"$BUN_CMD" run lint
echo "Running pre-commit Test..."
"$BUN_CMD" run test
```

Now i can open Fork normally via Finder/Spotlight/Dock :)

**Comment by CyberMew** [comment]
This feature has been requested before:

* https://github.com/fork-dev/Tracker/issues/2146
* https://github.com/fork-dev/Tracker/issues/430

It would be very helpful if Fork allowed configuring environment variables that are loaded when the app starts.

In my workflow, Fork launches automatically at system startup. Because of that, the current workaround of launching Fork from the terminal in order to inject environment variables is inconvenient, especially after restarts.

Other Git clients already support something similar. For example, Tower allows defining environment variables through a configuration file that the app loads on startup:
https://www.git-tower.com/help/guides/integration/environment/mac

My specific use case is Claude Code, where I bring my own endpoint and API key. These credentials must be provided through environment variables rather than stored login credentials.

Even a simple implementation would already help, for example:

• Loading variables from a `.env` file
• Or a small UI section where users can define environment variables that Fork exports when launching tools

Many Git clients simply inherit the system environment and do not provide a way to configure this directly. Adding this capability in Fork could make it especially convenient for developers integrating external tools, scripts, and AI workflows.

Being able to define this configuration directly in Fork would make these integrations much smoother and could become a nice differentiating feature for the app.

Thanks for considering it. Hope it can be added one day soon.

**Comment by kjyv** [comment]
@DanPristupov I see you're updating Fork regularly, great! But here I am still restarting Fork via `open -a Fork` from the terminal because of this issue every time after an update was installed... Could you pretty please consider including the same env settings that already happen for pre-commit also for pre-push?
